pulp_2to3_migration/app/migration.py

```diff
@@ -106,216 +143,194 @@ async def migrate_repositories(plan):
+            # re-run, a repo can be changed in pulp 2 and it might not be for the last
+            # repo version)
+            create_repo_version(plugin.migrator, repo_name, pulp2_repo, pulp3_remote)
+
+    for pulp2_repo_info in repo_versions_setup:
+        # find pulp2repo by id
+        repo_id = pulp2_repo_info['repo_id']
+        dist_repositories = pulp2_repo_info['dist_repo_ids']
+
+        try:
+            migrated_repo = Pulp2Repository.objects.get(pulp2_repo_id=repo_id,
+                                                        not_in_plan=False)
+        except Pulp2Repository.DoesNotExist:
+            # not in Pulp 2 anymore
+            continue
+        else:
+            pulp2dist = Pulp2Distributor.objects.filter(
+                pulp2_repo_id__in=dist_repositories,
+                pulp2_type_id__in=distributor_types
+            )
+            for dist in pulp2dist:
+                dist_migrator = distributor_migrators.get(dist.pulp2_type_id)
+                migrate_repo_distributor(
+                    dist_migrator, dist,
+                    migrated_repo.pulp3_repository_version
+                )
+
+
+async def create_repoversions_publications_distributions(plan, parallel=False):
     """
-    A coroutine to migrate pre-migrated importers.
+    A coroutine to create repository versions.
+
+    Content to a repo version is added based on pre-migrated RepoContentUnit and info provided
+    in the migration plan.
 
     Args:
         plan (MigrationPlan): Migration Plan to use.
-    """
-    # gather all needed plugin importer migrators
-    importer_migrators = {}
 
+    Kwargs:
+        parallel (bool): If True, attempt to migrate things in parallel where possible.
+    """
     for plugin in plan.get_plugin_plans():
-        importer_migrators.update(**plugin.migrator.importer_migrators)
+        pulp3_repo_setup = plugin.get_repo_creation_setup()
 
-    progress_data = dict(
-        message='Migrating importers to Pulp 3', code='migrating.importers', total=0
-    )
-    with ProgressReport(**progress_data) as pb:
-        pulp2importers_qs = Pulp2Importer.objects.filter(
-            pulp3_remote=None,
-            not_in_plan=False)
-        pb.total += pulp2importers_qs.count()
-        pb.save()
+        if not pulp3_repo_setup:
+            task_func = simple_plugin_migration
+            task_args = [plugin]
+            task_func(*task_args)
+        else:
+            task_func = complex_repo_migration
 
-        for pulp2importer in pulp2importers_qs:
-            importer_migrator = importer_migrators.get(pulp2importer.pulp2_type_id)
-            remote, created = await importer_migrator.migrate_to_pulp3(pulp2importer)
-            pulp2importer.pulp3_remote = remote
-            pulp2importer.is_migrated = True
-            pulp2importer.save()
-            if created:
-                pb.increment()
+            if parallel:
```

Comment:
The rest is just refactoring, but the meat of the PR is basically just these 2 dozen lines.  This threadpool code is the experiment I was talking about that only took about an hour to research and implement.

I did some not very scientific benchmarking to get an idea whether it might be nearly as fast or faster, here are the results:

Thread Pool

    Intial migration: 243 seconds
    Re-migration: 94 seconds

Multiple Tasks

    Intial migration: 245 seconds
    Re-migration: 94 seconds

Serialized (no parallel)

    Intial migration: 258 seconds
    Re-migration: 101 seconds

Reply:
Don't read too far into those results though, like I said, it was not very scientific.  I used Centos 6, 7, and 8-baseos, but I also had Fedora 30, 31, and Centos 8-appstream synced into Pulp 2 so these numbers are migrating all existing content from all 6 repos + those 3 repos I mentioned.  I gave the VM 5 CPUs, 12gb of memory, and ran 4 Pulp workers.  I also ran ```htop``` and I noticed that even with 12gb of memory and only 3 repos, it was still using swap space, which might mess with the numbers.